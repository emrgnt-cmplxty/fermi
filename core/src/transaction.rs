extern crate engine;
use serde::{Deserialize, Serialize};
use std::{fmt::Debug};

use diem_crypto::{
    Signature,
    hash::{CryptoHash, HashValue},
};
use diem_crypto_derive::{BCSCryptoHash, CryptoHasher};
use engine::orders::{OrderRequest};
use types::{
    account::{AccountPubKey, AccountSignature},
    asset::{AssetId},
    spot::{DiemCryptoMessage},
};

#[derive(BCSCryptoHash, Copy, Clone, CryptoHasher, Debug, Deserialize, Serialize)]
pub struct Payment 
{
    // storing from here is not redundant as from may not equal sender
    // e.g. we are preserving the possibility of adding re-key functionality
    from: AccountPubKey,
    to: AccountPubKey,
    asset_id: AssetId,
    amount: u64,
}
impl Payment {
    pub fn new(from: AccountPubKey, to: AccountPubKey, asset_id: AssetId, amount: u64) -> Self {
        Payment {
            from,
            to,
            asset_id,
            amount
        }
    }

    pub fn get_from(&self) -> &AccountPubKey {
        &self.from
    }

    pub fn get_to(&self) -> &AccountPubKey {
        &self.to
    }

    pub fn get_asset_id(&self) -> AssetId {
        self.asset_id
    }

    pub fn get_amount(&self) -> u64 {
        self.amount
    }
}
#[derive(BCSCryptoHash, Copy, Clone, CryptoHasher, Debug, Deserialize, Serialize)]
pub struct CreateAsset 
{
}

#[derive(BCSCryptoHash, Copy, Clone, CryptoHasher, Debug, Deserialize, Serialize)]
pub struct Order 
{
    order_request: OrderRequest,
}
impl Order {
    pub fn new(order_request: OrderRequest) -> Self {
        Order {
            order_request,
        }
    }

    pub fn get_order_request(&self) -> &OrderRequest {
        &self.order_request
    }
}

#[derive(BCSCryptoHash, Copy, Clone, CryptoHasher, Debug, Deserialize, Serialize)]
pub struct Stake 
{
    from: AccountPubKey,
    amount: u64,
}
impl Stake {
    pub fn new(from: AccountPubKey, amount: u64) -> Self {
        Stake {
            from,
            amount
        }
    }

    pub fn get_from(&self) -> &AccountPubKey {
        &self.from
    }

    pub fn get_amount(&self) -> u64 {
        self.amount
    }
}
#[derive(BCSCryptoHash, Copy, Clone, CryptoHasher, Debug, Deserialize, Serialize)]
pub enum TxnVariant {
    PaymentTransaction(Payment),
    CreateAssetTransaction(CreateAsset),
    OrderTransaction(Order),
    StakeAssetTransaction(Stake),
}
pub struct TxnRequest <TxnVariant>
where
    TxnVariant: Debug + Clone + CryptoHash + Copy,
{
    txn: TxnVariant,
    sender: AccountPubKey,
    txn_signature: AccountSignature,
}
impl <TxnVariant> TxnRequest <TxnVariant>
where
    TxnVariant: Debug + Clone + CryptoHash + Copy,
{
    pub fn new(txn: TxnVariant, sender: AccountPubKey, txn_signature: AccountSignature) -> Self {
        TxnRequest {
            txn,
            sender,
            txn_signature
        }
    }

    pub fn get_txn(&self) -> &TxnVariant {
        &self.txn
    }

    pub fn get_sender(&self) -> &AccountPubKey {
        &self.sender
    }

    pub fn get_txn_signature(&self) -> &AccountSignature {
        &self.txn_signature
    }

    pub fn verify_transaction(&self) -> Result<(), diem_crypto::error::Error> {
        let txn_hash: HashValue = self.txn.hash();
        self.txn_signature.verify(&DiemCryptoMessage(txn_hash.to_string()), &self.sender)
    }
}

