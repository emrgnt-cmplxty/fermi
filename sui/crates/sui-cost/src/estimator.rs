// Copyright (c) Mysten Labs, Inc.
// SPDX-License-Identifier: Apache-2.0

use anyhow::anyhow;
use serde::{Deserialize, Serialize};
use std::sync::Arc;
use strum_macros::Display;
use strum_macros::EnumString;
use sui_core::authority::AuthorityState;
use sui_core::test_utils::to_sender_signed_transaction;
use sui_core::transaction_input_checker;
use sui_types::base_types::ObjectID;
use sui_types::base_types::SequenceNumber;
use sui_types::base_types::TransactionDigest;
use sui_types::crypto::get_key_pair;
use sui_types::crypto::AccountKeyPair;
use sui_types::error::SuiResult;
use sui_types::gas::start_gas_metering;
use sui_types::gas::GasCostSummary;
use sui_types::gas::SuiGas;
use sui_types::gas::MAX_GAS_BUDGET;
use sui_types::gas_coin::GasCoin;
use sui_types::messages::SingleTransactionKind;
use sui_types::messages::TransactionData;
use sui_types::messages::TransactionKind;
use sui_types::temporary_store::TemporaryStore;

const DEFAULT_COMPUTATION_GAS_UNIT_PRICE: u64 = 1;
const DEFAULT_STORAGE_GAS_UNIT_PRICE: u64 = 1;
const DEFAULT_STORAGE_REBATE: u64 = 0;

#[derive(
    Debug, Eq, PartialEq, Hash, Serialize, Deserialize, Ord, PartialOrd, Clone, Display, EnumString,
)]
pub enum CommonTransactionCosts {
    Publish,
    MergeCoin,
    SplitCoin(usize),
    TransferWholeCoin,
    TransferWholeSuiCoin,
    TransferPortionSuiCoin,
    SharedCounterCreate,
    SharedCounterAssertValue,
    SharedCounterIncrement,
}

impl CommonTransactionCosts {
    pub fn is_shared_object_tx(&self) -> bool {
        matches!(
            self,
            CommonTransactionCosts::SharedCounterAssertValue
                | CommonTransactionCosts::SharedCounterIncrement
        )
    }
}

// Emulation-based estimator
// Step 1: charge min tx fee : can be computed precisely
// Step 2: if contains shared objs, charge consensus fee : can be computed precisely
// Step 3: charge storage read for all input objects : can be computed precisely
// Step 4: if package publish, charge for it per size of modules : can be computed precisely
// Step 5: charge VM flat fee if uses VM : can be computed precisely
// Step 6: charge for mutations, deletions, rebates: cannot be computed precisely, will approx
fn estimate_transaction_inner<S>(
    tx: TransactionKind,
    computation_gas_unit_price: Option<u64>,
    storage_gas_unit_price: Option<u64>,
    mutated_object_sizes_after: Option<usize>,
    storage_rebate: SuiGas,
    temporary_store: &TemporaryStore<S>,
) -> SuiResult<GasCostSummary> {
    let computation_gas_unit_price =
        computation_gas_unit_price.unwrap_or(DEFAULT_COMPUTATION_GAS_UNIT_PRICE);
    let storage_gas_unit_price = storage_gas_unit_price.unwrap_or(DEFAULT_STORAGE_GAS_UNIT_PRICE);

    let mut gas_status = start_gas_metering(
        *MAX_GAS_BUDGET,
        computation_gas_unit_price,
        storage_gas_unit_price,
    )?;

    // Step 1: charge min tx fee : can be computed precisely
    gas_status.charge_min_tx_gas()?;
    // Step 2: if contains shared objs, charge consensus fee : can be computed precisely
    if tx.shared_input_objects().next().is_some() {
        gas_status.charge_consensus()?;
    }
    // Step 3: charge storage read for all input objects : can be computed precisely
    // Dummy gas obj
    let gas_obj = GasCoin::new(ObjectID::random(), 0).to_object(SequenceNumber::new());
    let mut total_input_obj_size: usize = temporary_store
        .objects()
        .values()
        .map(|obj| obj.object_size_for_gas_metering())
        .sum();
    total_input_obj_size += gas_obj.object_size_for_gas_metering();
    gas_status.charge_storage_read(total_input_obj_size)?;

    // Steps 4 & 5
    for single_tx in tx.single_transactions() {
        match single_tx {
            SingleTransactionKind::Publish(module) => {
                gas_status.charge_publish_package(module.modules.iter().map(|v| v.len()).sum())?
            }
            SingleTransactionKind::Call(_) => (),
            _ => continue,
        }

        // Charge for Call and Publish
        // Emulate charging for flat fee, pending: https://github.com/MystenLabs/sui/pull/4607
        gas_status.charge_vm_gas()?;
    }

    // The assumption here is that size number of output objects is roughly double at most
    // This assumption is not based on any real data. Feedback is welcomed
    let mutated_object_sizes_after = mutated_object_sizes_after.unwrap_or(total_input_obj_size * 2);

    // Step 6: charge for mutations, deletions, rebates: cannot be computed precisely, will approx
    // At this point we need to estimate the effects of mutating objects after execution
    // We need to use some approx this
    gas_status.charge_storage_mutation(
        total_input_obj_size,
        mutated_object_sizes_after,
        storage_rebate,
    )?;

    Ok(gas_status.summary(true))
}

pub async fn estimate_transaction_computation_cost(
    tx_data: TransactionData,
    state: Arc<AuthorityState>,
    computation_gas_unit_price: Option<u64>,
    storage_gas_unit_price: Option<u64>,
    mutated_object_sizes_after: Option<usize>,
    storage_rebate: Option<u64>,
) -> anyhow::Result<GasCostSummary> {
    // Make a dummy transaction
    let (_, keypair): (_, AccountKeyPair) = get_key_pair();
    let tx = to_sender_signed_transaction(tx_data, &keypair);

    let (_gas_status, input_objects) =
        transaction_input_checker::check_transaction_input(&state.db(), &tx.data().data).await?;
    let in_mem_temporary_store =
        TemporaryStore::new(state.db(), input_objects, TransactionDigest::random());

    estimate_transaction_inner(
        tx.into_inner().into_data().data.kind,
        computation_gas_unit_price,
        storage_gas_unit_price,
        mutated_object_sizes_after,
        SuiGas::new(storage_rebate.unwrap_or(DEFAULT_STORAGE_REBATE)),
        &in_mem_temporary_store,
    )
    .map_err(|e| anyhow!("{e}"))
}
