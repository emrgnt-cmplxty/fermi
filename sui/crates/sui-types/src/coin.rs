// Copyright (c) Mysten Labs, Inc.
// SPDX-License-Identifier: Apache-2.0

use move_core_types::{
    ident_str,
    identifier::IdentStr,
    language_storage::{StructTag, TypeTag},
    value::{MoveFieldLayout, MoveStructLayout, MoveTypeLayout},
};
use serde::{Deserialize, Serialize};

use crate::base_types::TransactionDigest;
use crate::object::{MoveObject, Owner, OBJECT_START_VERSION};
use crate::storage::{DeleteKind, SingleTxContext, WriteKind};
use crate::temporary_store::TemporaryStore;
use crate::{
    balance::{Balance, Supply},
    error::{ExecutionError, ExecutionErrorKind},
    object::{Data, Object},
};
use crate::{
    base_types::{ObjectID, SuiAddress},
    id::UID,
    SUI_FRAMEWORK_ADDRESS,
};
use schemars::JsonSchema;

pub const COIN_MODULE_NAME: &IdentStr = ident_str!("coin");
pub const COIN_STRUCT_NAME: &IdentStr = ident_str!("Coin");

pub const PAY_MODULE_NAME: &IdentStr = ident_str!("pay");
pub const PAY_JOIN_FUNC_NAME: &IdentStr = ident_str!("join");
pub const PAY_SPLIT_N_FUNC_NAME: &IdentStr = ident_str!("divide_and_keep");
pub const PAY_SPLIT_VEC_FUNC_NAME: &IdentStr = ident_str!("split_vec");

// Rust version of the Move sui::coin::Coin type
#[derive(Debug, Serialize, Deserialize, Clone, JsonSchema, Eq, PartialEq)]
pub struct Coin {
    pub id: UID,
    pub balance: Balance,
}

impl Coin {
    pub fn new(id: UID, value: u64) -> Self {
        Self {
            id,
            balance: Balance::new(value),
        }
    }

    pub fn type_(type_param: StructTag) -> StructTag {
        StructTag {
            address: SUI_FRAMEWORK_ADDRESS,
            name: COIN_STRUCT_NAME.to_owned(),
            module: COIN_MODULE_NAME.to_owned(),
            type_params: vec![TypeTag::Struct(type_param)],
        }
    }

    /// Is this other StructTag representing a Coin?
    pub fn is_coin(other: &StructTag) -> bool {
        other.module.as_ident_str() == COIN_MODULE_NAME
            && other.name.as_ident_str() == COIN_STRUCT_NAME
    }

    /// Create a coin from BCS bytes
    pub fn from_bcs_bytes(content: &[u8]) -> Result<Self, ExecutionError> {
        bcs::from_bytes(content).map_err(|err| {
            ExecutionError::new_with_source(
                ExecutionErrorKind::InvalidCoinObject,
                format!("Unable to deserialize coin object: {:?}", err),
            )
        })
    }

    /// If the given object is a Coin, deserialize its contents and extract the balance Ok(Some(u64)).
    /// If it's not a Coin, return Ok(None).
    /// The cost is 2 comparisons if not a coin, and deserialization if its a Coin.
    pub fn extract_balance_if_coin(object: &Object) -> Result<Option<u64>, ExecutionError> {
        match &object.data {
            Data::Move(move_obj) => {
                if !Self::is_coin(&move_obj.type_) {
                    return Ok(None);
                }

                let coin = Self::from_bcs_bytes(move_obj.contents())?;
                Ok(Some(coin.value()))
            }
            _ => Ok(None), // package
        }
    }

    pub fn id(&self) -> &ObjectID {
        self.id.object_id()
    }

    pub fn value(&self) -> u64 {
        self.balance.value()
    }

    pub fn to_bcs_bytes(&self) -> Vec<u8> {
        bcs::to_bytes(&self).unwrap()
    }

    pub fn layout(type_param: StructTag) -> MoveStructLayout {
        MoveStructLayout::WithTypes {
            type_: Self::type_(type_param.clone()),
            fields: vec![
                MoveFieldLayout::new(
                    ident_str!("id").to_owned(),
                    MoveTypeLayout::Struct(UID::layout()),
                ),
                MoveFieldLayout::new(
                    ident_str!("balance").to_owned(),
                    MoveTypeLayout::Struct(Balance::layout(type_param)),
                ),
            ],
        }
    }

    // Shift balance of coins_to_merge to this coin.
    // Related coin objects need to be updated in temporary_store to presist the changes,
    // including deleting the coin objects that have been merged.
    pub fn merge_coins(&mut self, coins_to_merge: &mut [Coin]) {
        let total_coins = coins_to_merge.iter().fold(0, |acc, c| acc + c.value());
        for coin in coins_to_merge.iter_mut() {
            // unwrap() is safe because balance value is the same as coin value
            coin.balance.withdraw(coin.value()).unwrap();
        }
        self.balance = Balance::new(self.value() + total_coins);
    }

    // Split amount out of this coin to a new coin.
    // Related coin objects need to be updated in temporary_store to presist the changes,
    // including creating the coin object related to the newly created coin.
    pub fn split_coin(&mut self, amount: u64, new_coin_id: UID) -> Result<Coin, ExecutionError> {
        self.balance.withdraw(amount)?;
        Ok(Coin::new(new_coin_id, amount))
    }
}

// Rust version of the Move sui::coin::TreasuryCap type
#[derive(Debug, Serialize, Deserialize, Clone, Eq, PartialEq)]
pub struct TreasuryCap {
    pub id: UID,
    pub total_supply: Supply,
}

pub fn transfer_coin<S>(
    ctx: &SingleTxContext,
    temporary_store: &mut TemporaryStore<S>,
    coin: &Coin,
    recipient: SuiAddress,
    coin_type: StructTag,
    previous_transaction: TransactionDigest,
) {
    let new_coin = Object::new_move(
        MoveObject::new_coin(
            coin_type,
            OBJECT_START_VERSION,
            bcs::to_bytes(coin).expect("Serializing coin value cannot fail"),
        ),
        Owner::AddressOwner(recipient),
        previous_transaction,
    );
    temporary_store.write_object(ctx, new_coin, WriteKind::Create);
}

// A helper function for pay_sui and pay_all_sui.
// It updates the gas_coin_obj based on the updated gas_coin, transfers gas_coin_obj to
// recipient when needed, and then deletes all other input coins other than gas_coin_obj.
pub fn update_input_coins<S>(
    ctx: &SingleTxContext,
    temporary_store: &mut TemporaryStore<S>,
    coin_objects: &mut Vec<Object>,
    gas_coin: &Coin,
    recipient: Option<SuiAddress>,
) {
    let mut gas_coin_obj = coin_objects.remove(0);
    // unwrap is safe because we checked that it was a coin object above.
    // update_contents_without_version_change b/c this is the gas coin,
    // whose version will be bumped upon gas payment.
    gas_coin_obj
        .data
        .try_as_move_mut()
        .unwrap()
        .update_contents_without_version_change(
            bcs::to_bytes(gas_coin).expect("Coin serialization should not fail"),
        );
    if let Some(recipient) = recipient {
        gas_coin_obj.transfer_without_version_change(recipient);
    }
    temporary_store.write_object(ctx, gas_coin_obj, WriteKind::Mutate);

    for coin_object in coin_objects.iter() {
        temporary_store.delete_object(
            ctx,
            &coin_object.id(),
            coin_object.version(),
            DeleteKind::Normal,
        )
    }
}
