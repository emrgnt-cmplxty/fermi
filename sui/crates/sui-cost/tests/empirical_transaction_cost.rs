// Copyright (c) Mysten Labs, Inc.
// SPDX-License-Identifier: Apache-2.0

use insta::assert_json_snapshot;
use std::{collections::BTreeMap, path::PathBuf};
use sui_config::NetworkConfig;
use sui_config::ValidatorInfo;
use sui_cost::estimator::estimate_transaction_computation_cost;
use sui_cost::estimator::CommonTransactionCosts;
use sui_types::base_types::SuiAddress;
use sui_types::coin::PAY_JOIN_FUNC_NAME;
use sui_types::coin::PAY_MODULE_NAME;
use sui_types::coin::PAY_SPLIT_VEC_FUNC_NAME;
use sui_types::crypto::AccountKeyPair;
use sui_types::messages::VerifiedTransaction;
use sui_types::object::Object;
use sui_types::{
    gas::GasCostSummary,
    messages::{CallArg, ExecutionStatus, ObjectArg},
};
use test_utils::messages::make_transfer_object_transaction;
use test_utils::messages::make_transfer_sui_transaction;
use test_utils::messages::move_transaction_with_type_tags;
use test_utils::test_account_keys;
use test_utils::transaction::get_framework_object;
use test_utils::transaction::make_publish_package;
use test_utils::{
    authority::{spawn_test_authorities, test_authority_configs},
    messages::move_transaction,
    objects::test_gas_objects,
    transaction::{
        publish_counter_package, submit_shared_object_transaction, submit_single_owner_transaction,
    },
};

const TEST_DATA_DIR: &str = "tests/data/";

// Execute every entry function in Move framework and examples and ensure costs don't change
// To review snapshot changes, and fix snapshot differences,
// 0. Install cargo-insta
// 1. Run `cargo insta test --review` under `./sui-cost`.
// 2. Review, accept or reject changes.

#[tokio::test]
async fn test_good_snapshot() -> Result<(), anyhow::Error> {
    let mut common_costs_actual: BTreeMap<String, GasCostSummary> = BTreeMap::new();
    let mut common_costs_estimate: BTreeMap<String, GasCostSummary> = BTreeMap::new();

    run_actual_and_estimate_costs()
        .await?
        .iter()
        .for_each(|(k, (actual, estimate))| {
            common_costs_actual.insert(k.clone().to_string(), actual.clone());
            common_costs_estimate.insert(k.clone().to_string(), estimate.clone());
        });
    assert_json_snapshot!(common_costs_actual);
    assert_json_snapshot!(common_costs_estimate);

    Ok(())
}

async fn split_n_tx(
    n: u64,
    coin: &Object,
    gas: &Object,
    validator_info: &[ValidatorInfo],
) -> VerifiedTransaction {
    let split_amounts = vec![10u64; n as usize];
    let type_args = vec![coin.get_move_template_type().unwrap()];

    move_transaction_with_type_tags(
        gas.clone(),
        PAY_MODULE_NAME.as_str(),
        PAY_SPLIT_VEC_FUNC_NAME.as_str(),
        get_framework_object(validator_info)
            .await
            .compute_object_reference(),
        &type_args,
        vec![
            CallArg::Object(ObjectArg::ImmOrOwnedObject(coin.compute_object_reference())),
            CallArg::Pure(bcs::to_bytes(&split_amounts).unwrap()),
        ],
    )
}

async fn create_txes(
    sender: SuiAddress,
    keypair: &AccountKeyPair,
    gas_objects: &[Object],
    configs: &NetworkConfig,
) -> BTreeMap<CommonTransactionCosts, VerifiedTransaction> {
    let mut ret = BTreeMap::new();
    let mut gas_objects = gas_objects.to_vec().clone();
    // let _handles = spawn_test_authorities(gas_objects.clone(), configs).await;
    // Publish the move package to all authorities and get the new package ref.
    tokio::task::yield_now().await;
    tokio::time::sleep(std::time::Duration::from_secs(3)).await;

    //
    // Publish
    //
    let mut package_path = PathBuf::from(TEST_DATA_DIR);
    package_path.push("dummy_modules_publish");
    let publish_tx = make_publish_package(gas_objects.pop().unwrap(), package_path);
    ret.insert(CommonTransactionCosts::Publish, publish_tx);

    //
    // Transfer Whole Sui Coin and Transfer Portion of Sui Coin
    //
    let whole_sui_coin_tx = make_transfer_sui_transaction(
        gas_objects.pop().unwrap().compute_object_reference(),
        SuiAddress::default(),
        None,
        sender,
        keypair,
    );
    let partial_sui_coin_tx = make_transfer_sui_transaction(
        gas_objects.pop().unwrap().compute_object_reference(),
        SuiAddress::default(),
        Some(100),
        sender,
        keypair,
    );
    ret.insert(
        CommonTransactionCosts::TransferWholeSuiCoin,
        whole_sui_coin_tx,
    );
    ret.insert(
        CommonTransactionCosts::TransferPortionSuiCoin,
        partial_sui_coin_tx,
    );

    //
    // Transfer Whole Coin Object
    //
    let whole_coin_tx = make_transfer_object_transaction(
        gas_objects.pop().unwrap().compute_object_reference(),
        gas_objects.pop().unwrap().compute_object_reference(),
        sender,
        keypair,
        SuiAddress::default(),
    );

    ret.insert(CommonTransactionCosts::TransferWholeCoin, whole_coin_tx);

    //
    // Merge Two Coins
    //
    let c1 = gas_objects.pop().unwrap();
    let type_args = vec![c1.get_move_template_type().unwrap()];

    let merge_tx = move_transaction_with_type_tags(
        gas_objects.pop().unwrap(),
        PAY_MODULE_NAME.as_str(),
        PAY_JOIN_FUNC_NAME.as_str(),
        get_framework_object(configs.validator_set())
            .await
            .compute_object_reference(),
        &type_args,
        vec![
            CallArg::Object(ObjectArg::ImmOrOwnedObject(c1.compute_object_reference())),
            CallArg::Object(ObjectArg::ImmOrOwnedObject(
                gas_objects.pop().unwrap().compute_object_reference(),
            )),
        ],
    );
    ret.insert(CommonTransactionCosts::MergeCoin, merge_tx);

    //
    // Splt A Coin Into N Specific Amounts
    // Note spltting complexity does not depend on the amounts but only on the number of amounts
    //
    for n in 0..4 {
        let gas = gas_objects.pop().unwrap();
        let coin = gas_objects.pop().unwrap();
        let split_tx = split_n_tx(n, &gas, &coin, configs.validator_set())
            .await
            .clone();
        ret.insert(CommonTransactionCosts::SplitCoin(n as usize), split_tx);
    }

    //
    // Shared Object Section
    // Using the `counter` example
    //

    let package_ref =
        publish_counter_package(gas_objects.pop().unwrap(), configs.validator_set()).await;

    // Make a transaction to create a counter.
    tokio::task::yield_now().await;
    let transaction = move_transaction(
        gas_objects.pop().unwrap(),
        "counter",
        "create",
        package_ref,
        /* arguments */ Vec::default(),
    );
    let effects =
        submit_single_owner_transaction(transaction.clone(), configs.validator_set()).await;
    assert!(matches!(effects.status, ExecutionStatus::Success { .. }));
    let ((counter_id, counter_initial_shared_version, _), _) = effects.created[0];
    let counter_object_arg = ObjectArg::SharedObject {
        id: counter_id,
        initial_shared_version: counter_initial_shared_version,
    };

    ret.insert(CommonTransactionCosts::SharedCounterCreate, transaction);

    // Ensure the value of the counter is `0`.
    tokio::task::yield_now().await;
    let transaction = move_transaction(
        gas_objects.pop().unwrap(),
        "counter",
        "assert_value",
        package_ref,
        vec![
            CallArg::Object(counter_object_arg),
            CallArg::Pure(0u64.to_le_bytes().to_vec()),
        ],
    );

    ret.insert(
        CommonTransactionCosts::SharedCounterAssertValue,
        transaction,
    );

    // Make a transaction to increment the counter.
    tokio::task::yield_now().await;
    let transaction = move_transaction(
        gas_objects.pop().unwrap(),
        "counter",
        "increment",
        package_ref,
        vec![CallArg::Object(counter_object_arg)],
    );

    ret.insert(CommonTransactionCosts::SharedCounterIncrement, transaction);

    ret
}

async fn run_actual_and_estimate_costs(
) -> Result<BTreeMap<CommonTransactionCosts, (GasCostSummary, GasCostSummary)>, anyhow::Error> {
    let mut ret = BTreeMap::new();
    let gas_objects = test_gas_objects();
    let (sender, keypair) = test_account_keys().pop().unwrap();

    // Get the authority configs and spawn them. Note that it is important to not drop
    // the handles (or the authorities will stop).
    let configs = test_authority_configs();
    let handles = spawn_test_authorities(gas_objects.clone(), &configs).await;
    // Publish the move package to all authorities and get the new package ref.
    tokio::task::yield_now().await;
    tokio::time::sleep(std::time::Duration::from_secs(3)).await;

    let tx_map = create_txes(sender, &keypair, &gas_objects, &configs).await;

    for (tx_type, tx) in tx_map {
        let gas_used = if tx_type.is_shared_object_tx() {
            submit_shared_object_transaction(tx.clone(), configs.validator_set())
                .await
                .unwrap()
                .gas_cost_summary()
                .clone()
        } else {
            submit_single_owner_transaction(tx.clone(), configs.validator_set())
                .await
                .gas_cost_summary()
                .clone()
        };

        let gas_estimate = handles[0]
            .with_async(|node| async move {
                let state = node.state();
                estimate_transaction_computation_cost(
                    tx.into_inner().into_data().data,
                    state.clone(),
                    None,
                    None,
                    None,
                    None,
                )
                .await
                .unwrap()
            })
            .await;
        ret.insert(tx_type.clone(), (gas_used.clone(), gas_estimate.clone()));

        // Check that the estimates are not less than actual
        assert!(gas_used.computation_cost <= gas_estimate.computation_cost);
        assert!(gas_used.storage_cost <= gas_estimate.storage_cost);
        // Rebate should be less since this is returned
        assert!(gas_used.storage_rebate >= gas_estimate.storage_rebate);
    }
    Ok(ret)
}
