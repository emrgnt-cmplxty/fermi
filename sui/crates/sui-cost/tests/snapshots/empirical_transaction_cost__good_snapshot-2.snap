---
source: crates/sui-cost/tests/empirical_transaction_cost.rs
expression: common_costs_estimate
---
{
  "MergeCoin": {
    "computation_cost": 5792,
    "storage_cost": 8550,
    "storage_rebate": 0
  },
  "Publish": {
    "computation_cost": 5781,
    "storage_cost": 8485,
    "storage_rebate": 0
  },
  "SharedCounterAssertValue": {
    "computation_cost": 683,
    "storage_cost": 833,
    "storage_rebate": 0
  },
  "SharedCounterCreate": {
    "computation_cost": 562,
    "storage_cost": 802,
    "storage_rebate": 0
  },
  "SharedCounterIncrement": {
    "computation_cost": 683,
    "storage_cost": 833,
    "storage_rebate": 0
  },
  "SplitCoin": {
    "computation_cost": 5770,
    "storage_cost": 8518,
    "storage_rebate": 0
  },
  "TransferPortionSuiCoin": {
    "computation_cost": 56,
    "storage_cost": 52,
    "storage_rebate": 0
  },
  "TransferWholeCoin": {
    "computation_cost": 77,
    "storage_cost": 84,
    "storage_rebate": 0
  },
  "TransferWholeSuiCoin": {
    "computation_cost": 56,
    "storage_cost": 52,
    "storage_rebate": 0
  }
}
